File: wrappers/typescript/src/run-output-parser.ts
/**
 * run-output-parser.ts — parse the Mode A v2 subprocess outcome into a DisplayEvent.
 *
 * Implements §4.1 envelope schema and §4.4 (SC-D) precedence rules from
 * `docs/designs/2026-05-24-aaa-v2-mode-a-pivot-amendment.md`:
 *
 *   Rule 1 — envelope parseable per §4.1 → envelope is authoritative.
 *     The `error` field (null or populated) drives the wrapper's outcome.
 *     The exit code is informational and does NOT override the envelope.
 *
 *   Rule 2 — envelope absent / unparseable / partial → synthesize an error
 *            event from exit code and stderr tail. Partial JSON is NOT
 *            half-parsed (belt-and-suspenders): if any required §4.1 field
 *            is missing, the envelope is treated as unparseable.
 *
 * stderrTail is truncated to STDERR_TAIL_BYTES (4096) on synthesized paths;
 * on the envelope path it is taken verbatim from the engine.
 */

import type { DisplayEvent } from "./session.js";

/** Maximum stderrTail length retained on synthesized engine errors. */
export const STDERR_TAIL_BYTES = 4096;

/** Maximum stdout snippet included in `envelope_missing` messages. */
const STDOUT_PREVIEW_BYTES = 512;

/** Outcome of running the `amplifier-agent run --output json` subprocess. */
export interface SubprocessOutcome {
  stdout: string;
  stderr: string;
  exitCode: number;
}

/**
 * Keep the last `STDERR_TAIL_BYTES` chars of `stderr`.
 * Returns `undefined` for an empty string so callers can omit the field
 * cleanly when there is nothing to surface.
 */
function tailStderr(stderr: string): string | undefined {
  if (!stderr) return undefined;
  if (stderr.length <= STDERR_TAIL_BYTES) return stderr;
  return stderr.slice(stderr.length - STDERR_TAIL_BYTES);
}

/** Allowed values for `error.classification` per §4.1. */
type Classification = "transport" | "protocol" | "engine" | "approval" | "unknown";
const VALID_CLASSIFICATIONS: ReadonlySet<string> = new Set<string>([
  "transport",
  "protocol",
  "engine",
  "approval",
  "unknown",
]);

/**
 * Validate that `parsed` conforms to the §4.1 envelope shape.
 *
 * Required:
 *   - protocolVersion, sessionId, turnId, reply: string
 *   - error: null | object with `code: string`
 *   - metadata: object
 *
 * Partial / type-wrong envelopes return `false` so the caller falls to Rule 2.
 */
function isShapeValid(parsed: unknown): parsed is {
  protocolVersion: string;
  sessionId: string;
  turnId: string;
  reply: string;
  error: null | {
    code: string;
    classification?: string;
    severity?: string;
    correlationId?: string;
    message?: string;
    stderrTail?: string;
  };
  metadata: Record<string, unknown>;
} {
  if (parsed === null || typeof parsed !== "object") return false;
  const o = parsed as Record<string, unknown>;
  if (typeof o.protocolVersion !== "string") return false;
  if (typeof o.sessionId !== "string") return false;
  if (typeof o.turnId !== "string") return false;
  if (typeof o.reply !== "string") return false;
  if (typeof o.metadata !== "object" || o.metadata === null) return false;

  if (o.error === null) return true;
  if (typeof o.error !== "object") return false;
  const err = o.error as Record<string, unknown>;
  if (typeof err.code !== "string") return false;
  return true;
}

/**
 * Parse a subprocess outcome into a single DisplayEvent.
 *
 * See module docstring for precedence rules.
 */
export function parseRunOutput(outcome: SubprocessOutcome): DisplayEvent {
  const trimmed = outcome.stdout.trim();

  // Attempt to parse stdout as JSON. Failures (empty, partial, non-JSON) are
  // captured silently; the caller falls to Rule 2.
  let parsed: unknown = null;
  if (trimmed.length > 0) {
    try {
      parsed = JSON.parse(trimmed);
    } catch {
      parsed = null;
    }
  }

  // Rule 1 — envelope parseable per §4.1 → envelope wins.
  if (parsed !== null && isShapeValid(parsed)) {
    const env = parsed;

    if (env.error === null) {
      // Success path — exit code is informational only.
      return { type: "result", text: env.reply };
    }

    // Failure path — populate from the envelope's error fields.
    const err = env.error;
    const classification: Classification =
      err.classification !== undefined &&
      VALID_CLASSIFICATIONS.has(err.classification)
        ? (err.classification as Classification)
        : "unknown";
    const severity: "error" | "warning" =
      err.severity === "warning" ? "warning" : "error";
    const correlationId =
      typeof err.correlationId === "string" ? err.correlationId : "";
    const message = typeof err.message === "string" ? err.message : err.code;
    const stderrTail =
      typeof err.stderrTail === "string" ? err.stderrTail : tailStderr(outcome.stderr);

    return {
      type: "error",
      code: err.code,
      classification,
      severity,
      correlationId,
      message,
      ...(stderrTail !== undefined ? { stderrTail } : {}),
      retryable: false,
    };
  }

  // Rule 2 — envelope absent or unparseable → synthesize from exit + stderr.
  const stderrTail = tailStderr(outcome.stderr);

  if (outcome.exitCode === 0) {
    // Engine protocol violation: exit 0 without a parseable envelope.
    const preview = outcome.stdout.slice(0, STDOUT_PREVIEW_BYTES);
    const previewSuffix =
      outcome.stdout.length > STDOUT_PREVIEW_BYTES ? "...(truncated)" : "";
    return {
      type: "error",
      code: "envelope_missing",
      classification: "protocol",
      severity: "error",
      correlationId: "",
      message: `Engine exited 0 without emitting a parseable §4.1 envelope. Stdout was: ${JSON.stringify(preview)}${previewSuffix}`,
      ...(stderrTail !== undefined ? { stderrTail } : {}),
      retryable: false,
    };
  }

  // Non-zero exit, envelope absent or partial — engine-class failure.
  return {
    type: "error",
    code: `engine_exit_${outcome.exitCode}`,
    classification: "engine",
    severity: "error",
    correlationId: "",
    message: `Engine exited ${outcome.exitCode} without emitting a parseable §4.1 envelope.`,
    ...(stderrTail !== undefined ? { stderrTail } : {}),
    retryable: false,
  };
}

File: wrappers/python-py/src/amplifier_agent_py/run_output_parser.py
"""Parse the Mode A v2 subprocess outcome into a single ``DisplayEvent``.

Mirrors wrappers/typescript/src/run-output-parser.ts.

Implements §4.1 envelope schema and §4.4 (SC-D) precedence rules from
``docs/designs/2026-05-24-aaa-v2-mode-a-pivot-amendment.md``:

  Rule 1 — envelope parseable per §4.1 → envelope is authoritative.
    The ``error`` field (null or populated) drives the wrapper's outcome.
    The exit code is informational and does NOT override the envelope.

  Rule 2 — envelope absent / unparseable / partial → synthesize an error
           event from exit code and stderr tail.  Partial JSON is NOT
           half-parsed (belt-and-suspenders): if any required §4.1 field
           is missing, the envelope is treated as unparseable.

``stderr_tail`` is truncated to ``STDERR_TAIL_BYTES`` (4096) on synthesized
paths; on the envelope path it is taken verbatim from the engine.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, cast

from .errors import Classification
from .types import DisplayEvent, ErrorEvent, ResultEvent

#: Maximum stderr_tail length retained on synthesized engine errors.
STDERR_TAIL_BYTES = 4096

#: Maximum stdout snippet included in ``envelope_missing`` messages.
_STDOUT_PREVIEW_BYTES = 512

_VALID_CLASSIFICATIONS: frozenset[str] = frozenset({"transport", "protocol", "engine", "approval", "unknown"})


@dataclass(frozen=True, kw_only=True)
class SubprocessOutcome:
    """Outcome of running the ``amplifier-agent run --output json`` subprocess."""

    stdout: str
    stderr: str
    exit_code: int


def _tail_stderr(stderr: str) -> str | None:
    """Keep the last ``STDERR_TAIL_BYTES`` chars of ``stderr`` or ``None``."""
    if not stderr:
        return None
    if len(stderr) <= STDERR_TAIL_BYTES:
        return stderr
    return stderr[-STDERR_TAIL_BYTES:]


def _is_shape_valid(parsed: Any) -> bool:
    """Validate that ``parsed`` conforms to the §4.1 envelope shape."""
    if not isinstance(parsed, dict):
        return False
    if not isinstance(parsed.get("protocolVersion"), str):
        return False
    if not isinstance(parsed.get("sessionId"), str):
        return False
    if not isinstance(parsed.get("turnId"), str):
        return False
    if not isinstance(parsed.get("reply"), str):
        return False
    if not isinstance(parsed.get("metadata"), dict):
        return False

    err = parsed.get("error")
    if err is None:
        return True
    if not isinstance(err, dict):
        return False
    return isinstance(err.get("code"), str)


def parse_run_output(outcome: SubprocessOutcome) -> DisplayEvent:
    """Parse a subprocess outcome into a single ``DisplayEvent``.

    See module docstring for precedence rules.
    """
    trimmed = outcome.stdout.strip()

    parsed: Any = None
    if trimmed:
        try:
            parsed = json.loads(trimmed)
        except (json.JSONDecodeError, ValueError):
            parsed = None

    # Rule 1 — envelope parseable per §4.1 → envelope wins.
    if parsed is not None and _is_shape_valid(parsed):
        env = cast(dict[str, Any], parsed)

        err = env.get("error")
        if err is None:
            return ResultEvent(text=cast(str, env["reply"]))

        # Failure path — populate from the envelope's error fields.
        err_dict = cast(dict[str, Any], err)
        raw_class = err_dict.get("classification")
        classification: Classification = (
            cast(Classification, raw_class) if raw_class in _VALID_CLASSIFICATIONS else "unknown"
        )
        severity = "warning" if err_dict.get("severity") == "warning" else "error"
        correlation_id_raw = err_dict.get("correlationId")
        correlation_id = correlation_id_raw if isinstance(correlation_id_raw, str) else ""
        message_raw = err_dict.get("message")
        message = message_raw if isinstance(message_raw, str) else cast(str, err_dict["code"])

        envelope_tail = err_dict.get("stderrTail")
        if isinstance(envelope_tail, str):
            stderr_tail: str | None = envelope_tail
        else:
            stderr_tail = _tail_stderr(outcome.stderr)

        return ErrorEvent(
            code=cast(str, err_dict["code"]),
            classification=classification,
            severity=severity,
            correlation_id=correlation_id,
            message=message,
            stderr_tail=stderr_tail,
            retryable=False,
        )

    # Rule 2 — envelope absent or unparseable → synthesize from exit + stderr.
    stderr_tail = _tail_stderr(outcome.stderr)

    if outcome.exit_code == 0:
        preview = outcome.stdout[:_STDOUT_PREVIEW_BYTES]
        preview_suffix = "...(truncated)" if len(outcome.stdout) > _STDOUT_PREVIEW_BYTES else ""
        return ErrorEvent(
            code="envelope_missing",
            classification="protocol",
            severity="error",
            correlation_id="",
            message=(
                f"Engine exited 0 without emitting a parseable §4.1 envelope. "
                f"Stdout was: {json.dumps(preview)}{preview_suffix}"
            ),
            stderr_tail=stderr_tail,
            retryable=False,
        )

    return ErrorEvent(
        code=f"engine_exit_{outcome.exit_code}",
        classification="engine",
        severity="error",
        correlation_id="",
        message=f"Engine exited {outcome.exit_code} without emitting a parseable §4.1 envelope.",
        stderr_tail=stderr_tail,
        retryable=False,
    )
